crates/nu-path/src/pwd_per_drive.rs

```diff
@@ -0,0 +1,366 @@
+    }
+
+    /// Set the PWD for the drive letter in the absolute path.
+    /// Return String for error description.
+    pub fn set_pwd(&mut self, path: &Path) -> Result<(), String> {
+        if let (Some(drive_letter), Some(path_str)) = (
+            Self::extract_drive_letter(path),
+            path.to_str(),
+        ) {
+            self.map[drive_letter as usize - 'A' as usize] = Some(path_str.to_string());
+            return Ok(());
+        }
+        Err(format!("Invalid path: {}", path.display()))
+    }
+
+    /// Get the PWD for drive, if not yet, ask GetFullPathNameW(),
+    /// or else return default r"X:\".
+    fn get_pwd(&mut self, drive: char) -> Option<String> {
+        if drive.is_ascii_alphabetic() {
+            let drive = drive.to_ascii_uppercase();
+            let index = drive as usize - 'A' as usize;
+            Some(self.map[index]
+                .clone()
+                .unwrap_or_else(||
+                    if let Some(system_pwd) = get_full_path_name_w(&format!("{}:", drive)) {
+                        self.map[index] = Some(system_pwd.clone());
+                        system_pwd
+                    } else {
+                        format!(r"{}:\", drive)
+                    }
+                )
+            )
+        } else {
+            None
+        }
+    }
+
+    /// Expand a relative path using the PWD-per-drive, return PathBuf
+    /// of absolute path.
+    /// Return None if path is not valid or can't get drive letter.
+    pub fn expand_path(&mut self, path: &Path) -> Option<PathBuf> {
+        let path_str = path.to_str()?;
+        if let Some(drive_letter) = Self::extract_drive_letter(path) {
+            if let Some(pwd) = self.get_pwd(drive_letter) {
+                // Combine current PWD with the relative path
+                let mut base = PathBuf::from(Self::ensure_trailing_separator(&pwd));
+                base.push(path_str.split_at(2).1); // Skip the "C:" part of the relative path
+                return Some(base)
+            }
+        }
+        None // Invalid path or has no drive letter
+    }
+
+    /// Extract the drive letter from a path (e.g., `C:test` -> `C`)
+    fn extract_drive_letter(path: &Path) -> Option<char> {
+        Some(path.to_str()
+            .and_then(|s| s.chars().next())
+            .filter(|c| c.is_ascii_alphabetic())?
+        )
+    }
+
+    /// Ensure a path has a trailing `\`
+    fn ensure_trailing_separator(path: &str) -> String {
+        if !path.ends_with('\\') && !path.ends_with('/') {
+            format!(r"{}\", path)
+        } else {
+            path.to_string()
+        }
+    }
+}
+
+// GetFullPathW
+fn get_full_path_name_w(path_str: &str) -> Option<String> {
+    use std::ffi::OsString;
+    use std::os::windows::ffi::OsStringExt;
+    use std::os::windows::ffi::OsStrExt;
+    use winapi::um::fileapi::GetFullPathNameW;
+
+    const MAX_PATH : usize = 260;
+    let mut buffer: [u16; MAX_PATH] = [0; MAX_PATH];
+
+    unsafe {
```

Comment:
This unsafe is probably ok since you probably have to have it for GetFullPathNameW

Reply:
replace 'winapi' crate with 'path_abs', this segment of unsafe code eliminated.